File: DSA/Python/Postorder Traversal on Binary Tree (Recursive).py
# In this program we will perform Postorder traversal on a binary tree

class Node:
    def __init__(self, data):
        self.val = data
        self.left = None
        self.right = None

def build_tree():
    data = int(input("Enter data: "))
    
    if data == -1:
        return None
    
    root = Node(data)
    
    print(f"Enter data in left of {data}")
    root.left = build_tree()
    
    print(f"Enter data in right of {data}")
    root.right = build_tree()
    
    return root

def postorder(root):
    if root is None:
        return
    
    postorder(root.left)
    postorder(root.right)
    print(root.val, end=" ")

root = build_tree()
print("Postorder Traversal:", end=" ")
postorder(root)

File: Data Structures And Algorithm/Java/Postorder Traversal on Binary Tree (Recursive).java
// In this program we will perform Postorder traversal on a binary tree

import java.util.Scanner;

class Node {
    int val;
    Node left, right;

    Node(int data) {
        val = data;
        left = null;
        right = null;
    }
}

public class BinaryTree {

    static Node buildTree() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter data : ");
        int data = scanner.nextInt();

        if (data == -1) {
            return null;
        }

        Node root = new Node(data);

        System.out.println("Enter data in left of " + data);
        root.left = buildTree();

        System.out.println("Enter data in right of " + data);
        root.right = buildTree();

        return root;
    }

    static void postOrder(Node root) {
        if (root == null) {
            return;
        }

        postOrder(root.left);
        postOrder(root.right);
        System.out.print(root.val + " ");
    }

    public static void main(String[] args) {
        Node root = buildTree();
        System.out.print("Postorder Traversal: ");
        postOrder(root);
    }
}
